### ISSUE TITLE
`video.add_text.ffmpeg` expects `string?` text, not a callable `{string}` like other formats

### ISSUE THREAD

**Issue by SlvrEagle23** [savonet/liquidsoap#2050]
Hello all!

Just playing around with Liquidsoap 2.0.1, in particular trying to get a fully-Liquidsoap implementation of [AzuraCast's radio-video-stream shell script](https://github.com/AzuraCast/radio-video-stream/blob/main/videostream/broadcast.sh), only to run into one unfortunate stumbling block:

It appears that the API for `video.add_text.ffmpeg` is _slightly_ different from the other `video.add_text` functions in that instead of accepting a callable for its text that can be updated periodically, it only accepts a one-time string, making updating the string impossible after it's initially supplied.

The goal for our implementation is to write the now-playing metadata to the video stream as a line of text at the bottom of the screen; this is done by a huge number of YouTube/Twitch-based "radio stations" that are operating as video streams either primarily or as a secondary broadcast destination.

Is it possible to adapt ffmpeg to get it to function the way the other `video.add_text` implementations (GD, native, SDL, etc.) behave?

As an aside, for the interim, what would be the best tool to use out of the existing ones to accomplish this? I've tried using gstreamer but it runs into immediate issues trying to read the video file (the error in question is `[clock.ffmpeg.encode.video.producer_0.child:2] Source ffmpeg.encode.video.consumer_0 failed while streaming: Gstreamer.Error("no element \"videotestsrc\"")!`, so I'm not sure if that's gstreamer or ffmpeg or their interaction messing up); not sure if GD or SDL would offer a better experience, but I know the native implementation doesn't support custom fonts so that won't work for this situation.

Thanks all! :)

**Comment by toots** [comment]
Hi @SlvrEagle23! 

I was writing a whole response and went back to ffmpeg's documentation only to realize that there is now a notion of comands for filters. Looks like it is indeed possible to change the variables dynamically..! See for instance: https://stackoverflow.com/a/67726791 and https://ffmpeg.org/ffmpeg-filters.html#Commands-78.

I'll look at implementing that soon and will try to add it to `2.0.1` if it looks stable enough!

As for gstreamer, I would avoid using it whenever possible. The implementation is considered deprecated at this point, in favor of ffmpeg.

**Comment by toots** [comment]
It's here: https://github.com/savonet/liquidsoap/pull/2062 !

**Comment by SlvrEagle23** [comment]
Excellent, thank you!!

**Comment by SlvrEagle23** [comment]
@toots Is this actually implemented as of the 2.0.2 release?

I'm getting an error when I try to pass a callable string function to `video.add_text.ffmpeg`:

```
Error 5: this value has type
  () -> _
but it should be a subtype of the type of the value at /var/azuracast/.opam/4.12.0/lib/liquidsoap/share/liquidsoap/2.0.2/libs/ffmpeg.liq, line 154, char 26
  string? (inferred at /var/azuracast/.opam/4.12.0/lib/liquidsoap/share/liquidsoap/2.0.2/libs/ffmpeg.liq, line 89, char 56)
```

In a related hurdle, I'm getting:

```
Error 5: this value has type
  source(audio=none,...) (inferred at /var/azuracast/stations/azuratest_radio/config/liquidsoap.liq, line 268, char 14-44)
but it should be a subtype of the type of the value at /var/azuracast/.opam/4.12.0/lib/liquidsoap/share/liquidsoap/2.0.2/libs/ffmpeg.liq, line 154, char 29
  source(audio=pcm(_),...) (inferred at /var/azuracast/.opam/4.12.0/lib/liquidsoap/share/liquidsoap/2.0.2/libs/ffmpeg.liq, line 156, char 69)
At /var/azuracast/stations/azuratest_radio/config/liquidsoap.liq, line 280, char 4-15:
    videostream
```

When calling:

```
videostream = ffmpeg.encode.video(
    %ffmpeg(%video(codec="libx264", pixel_format="yuv420p", b="300k", preset="superfast", r=25, g=50)), 
    videostream
)
```

There's no reason that should be inferring that it needs a signal with audio, right?

Can report it as a separate issue, but trying to make sure that I'm not expecting these things on the wrong version.
